File: hyperkit-agent/scripts/ci/update_version_in_docs.py
#!/usr/bin/env python3
"""
Automated Version Injection Script
Syncs version numbers across all docs, reports, and markdown files

Note: This script updates files in both main and devlog branches.
Files in devlog-only directories (REPORTS/, docs/TEAM/, etc.) will be
synced to devlog branch on next sync.
"""

import os
import re
import glob
from pathlib import Path
from datetime import datetime

def get_version():
    """Get version from VERSION file (root directory only - source of truth)"""
    # ROOT DIRECTORY ONLY - Single source of truth
    script_dir = Path(__file__).parent.parent.parent
    root_version = script_dir.parent / "VERSION"  # repo root VERSION
    
    if not root_version.exists():
        print(f"ERROR: VERSION file not found in repo root: {root_version.resolve()}")
        return None
    
    try:
        with open(root_version, "r") as f:
            version = f.read().strip()
            print(f"📁 Using root VERSION file: {root_version.resolve()}")
            return version
    except Exception as e:
        print(f"ERROR reading VERSION file: {e}")
        return None

def get_git_info():
    """Get git commit hash and date"""
    try:
        import subprocess
        # Get current directory - script may run from hyperkit-agent/ or repo root
        script_dir = Path(__file__).parent.parent.parent
        original_dir = os.getcwd()
        try:
            os.chdir(script_dir)  # Change to repo root for git commands
            
            commit_hash = subprocess.check_output(["git", "rev-parse", "--short", "HEAD"], 
                                                text=True, stderr=subprocess.DEVNULL).strip()
            commit_date = subprocess.check_output(["git", "log", "-1", "--format=%Y-%m-%d"], 
                                                 text=True, stderr=subprocess.DEVNULL).strip()
            
            # Fallback if date parsing fails or is invalid
            if not commit_date or len(commit_date) == 0 or not commit_date[0].isdigit():
                commit_date = datetime.now().strftime("%Y-%m-%d")
            
            return commit_hash, commit_date
        finally:
            os.chdir(original_dir)  # Restore original directory
    except Exception as e:
        print(f"Warning: Could not get git info: {e}")
        return "unknown", datetime.now().strftime("%Y-%m-%d")

def update_version_in_file(file_path, version, commit_hash, commit_date):
    """Update version markers in a single file"""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        original_content = content
        
        # Update VERSION_PLACEHOLDER blocks
        version_pattern = r'<!-- VERSION_PLACEHOLDER -->.*?<!-- /VERSION_PLACEHOLDER -->'
        version_replacement = f'''<!-- VERSION_PLACEHOLDER -->
**Version**: {version}
**Last Updated**: {commit_date}
**Commit**: {commit_hash}
<!-- /VERSION_PLACEHOLDER -->'''
        
        content = re.sub(version_pattern, version_replacement, content, flags=re.DOTALL)
        
        # Update SCRIPT_MARKER blocks
        script_pattern = r'<!-- SCRIPT_MARKER name=([^ ]+) version=([^ ]+) hash=([^ ]+) -->'
        def update_script_marker(match):
            script_name = match.group(1)
            return f'<!-- SCRIPT_MARKER name={script_name} version={version} hash={commit_hash} -->'
        
        content = re.sub(script_pattern, update_script_marker, content)
        
        # Update standalone version references
        content = re.sub(r'\*\*Version\*\*:\s*\d+\.\d+\.\d+', f'**Version**: {version}', content)
        
        # Write back if changed
        if content != original_content:
            with open(file_path, 'w', encoding='utf-8') as f:
                f.write(content)
            print(f"Updated: {file_path}")
            return True
        
        return False
        
    except Exception as e:
        print(f"Error updating {file_path}: {e}")
        return False

def main():
    """Main function to update all markdown files"""
    version = get_version()
    if not version:
        return 1
    
    commit_hash, commit_date = get_git_info()
    
    print(f"Updating version to {version} (commit: {commit_hash}, date: {commit_date})")
    
    # Find all markdown files
    # Script runs from hyperkit-agent/, so we need to check parent directories too
    markdown_files = []
    
    # Patterns relative to current directory (hyperkit-agent/)
    patterns = [
        "**/*.md",                    # Current directory and subdirectories
        "docs/**/*.md",               # Local docs
        "REPORTS/**/*.md",            # Local reports
    ]
    
    # Patterns for parent directory (repo root) - SOURCE OF TRUTH
    parent_patterns = [
        "../README.md",               # Root README (source of truth)
        "../CHANGELOG.md",            # Root CHANGELOG (source of truth)
        "../SECURITY.md",             # Root SECURITY (source of truth)
        "../docs/**/*.md",            # Root docs if exists
    ]
    
    # Collect files from current directory
    for pattern in patterns:
        markdown_files.extend(glob.glob(pattern, recursive=True))
    
    # Collect files from parent directory
    for pattern in parent_patterns:
        found = glob.glob(pattern, recursive=True)
        markdown_files.extend(found)
    
    # Remove duplicates and sort
    markdown_files = sorted(set(markdown_files))
    
    updated_count = 0
    for file_path in markdown_files:
        if update_version_in_file(file_path, version, commit_hash, commit_date):
            updated_count += 1
    
    print(f"Updated {updated_count} files")
    return 0

if __name__ == "__main__":
    exit(main())

File: hyperkit-agent/scripts/release/update-version-in-docs.js
#!/usr/bin/env node
/**
 * Update Version in All Documentation
 * 
 * Syncs version number across all documentation files:
 * - README.md
 * - CHANGELOG.md
 * - All markdown files with version references
 * - Documentation headers and badges
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

const ROOT_DIR = path.resolve(__dirname, '../..');
const VERSION_FILE = path.join(ROOT_DIR, 'VERSION');

function getCurrentVersion() {
  try {
    if (fs.existsSync(VERSION_FILE)) {
      return fs.readFileSync(VERSION_FILE, 'utf8').trim();
    }
    const pkg = JSON.parse(fs.readFileSync(path.join(ROOT_DIR, 'package.json'), 'utf8'));
    return pkg.version;
  } catch (err) {
    console.error('❌ Could not read current version');
    process.exit(1);
  }
}

function findMarkdownFiles(dir, files = []) {
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  
  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    
    // Skip certain directories
    if (entry.isDirectory()) {
      if (['node_modules', '.git', '__pycache__', '.pytest_cache', 'dist', 'build'].includes(entry.name)) {
        continue;
      }
      findMarkdownFiles(fullPath, files);
    } else if (entry.isFile() && entry.name.endsWith('.md')) {
      files.push(fullPath);
    }
  }
  
  return files;
}

function updateVersionInFile(filePath, version) {
  let content = fs.readFileSync(filePath, 'utf8');
  let updated = false;
  
  // Pattern 1: Version badges [![Version](...vX.Y.Z...)]
  const badgePattern = /\[!\[Version\]\([^)]*v\d+\.\d+\.\d+[^)]*\)\]/g;
  if (badgePattern.test(content)) {
    content = content.replace(
      /v\d+\.\d+\.\d+/g,
      `v${version}`
    );
    updated = true;
  }
  
  // Pattern 2: Version: X.Y.Z or Version X.Y.Z
  const versionPattern = /(Version|version):\s*\d+\.\d+\.\d+/g;
  if (versionPattern.test(content)) {
    content = content.replace(
      /(Version|version):\s*\d+\.\d+\.\d+/g,
      `$1: ${version}`
    );
    updated = true;
  }
  
  // Pattern 3: vX.Y.Z in text
  const textVersionPattern = /\bv\d+\.\d+\.\d+\b/g;
  if (textVersionPattern.test(content)) {
    content = content.replace(
      /\bv\d+\.\d+\.\d+\b/g,
      `v${version}`
    );
    updated = true;
  }
  
  // Pattern 4: HyperKit Agent vX.Y.Z
  const agentVersionPattern = /HyperKit Agent v\d+\.\d+\.\d+/gi;
  if (agentVersionPattern.test(content)) {
    content = content.replace(
      /HyperKit Agent v\d+\.\d+\.\d+/gi,
      `HyperKit Agent v${version}`
    );
    updated = true;
  }
  
  if (updated) {
    fs.writeFileSync(filePath, content, 'utf8');
    return true;
  }
  
  return false;
}

function gitAdd(filePath) {
  try {
    const relativePath = path.relative(ROOT_DIR, filePath);
    execSync(`git add "${relativePath}"`, {
      cwd: ROOT_DIR,
      stdio: 'pipe'
    });
    return true;
  } catch (err) {
    return false;
  }
}

function gitCommit(files, message) {
  try {
    for (const file of files) {
      gitAdd(file);
    }
    execSync(`git commit -m "${message}"`, {
      cwd: ROOT_DIR,
      stdio: 'pipe'
    });
    return true;
  } catch (err) {
    if (err.message.includes('nothing to commit')) {
      return false;
    }
    return false;
  }
}

function main() {
  const autoCommit = !process.argv.includes('--no-commit');
  
  console.log(`\n📚 Updating version in all documentation files\n`);
  
  const version = getCurrentVersion();
  console.log(`Current version: ${version}\n`);
  
  // Find all markdown files
  const mdFiles = findMarkdownFiles(ROOT_DIR);
  console.log(`Found ${mdFiles.length} markdown files to check\n`);
  
  const updatedFiles = [];
  
  for (const filePath of mdFiles) {
    // Skip certain files
    const relativePath = path.relative(ROOT_DIR, filePath);
    if (relativePath.includes('node_modules') || relativePath.includes('.git')) {
      continue;
    }
    
    if (updateVersionInFile(filePath, version)) {
      updatedFiles.push(filePath);
      console.log(`✅ Updated: ${relativePath}`);
    }
  }
  
  if (updatedFiles.length === 0) {
    console.log(`ℹ️  No documentation files needed version updates`);
    return;
  }
  
  console.log(`\n✅ Updated ${updatedFiles.length} documentation file(s)`);
  
  if (autoCommit) {
    if (gitCommit(updatedFiles, `docs: update version references to ${version}`)) {
      console.log(`\n✅ Committed version updates`);
    } else {
      console.log(`\n⚠️  No changes to commit`);
    }
  } else {
    console.log(`\n📝 Next steps:`);
    console.log(`   1. Review changes: git diff`);
    console.log(`   2. Commit: git add ${updatedFiles.map(f => path.relative(ROOT_DIR, f)).join(' ')}`);
    console.log(`   3. Commit: git commit -m "docs: update version references to ${version}"`);
  }
  
  console.log();
}

if (require.main === module) {
  main();
}

module.exports = { getCurrentVersion, updateVersionInFile };

